word_language_model/model.py

```diff
@@ -55,3 +57,165 @@ def init_hidden(self, bsz):
+            src_mask: the mask for the src sequence (optional).
+            tgt_mask: the mask for the tgt sequence (optional).
+            memory_mask: the mask for the encoder output (optional).
+        Shape:
+            src: [source sequence length, batch size]
+            tgt: [target sequence length, batch size]
+            src_mask: [source sequence length, source sequence length]
+            tgt_mask: [target sequence length, target sequence length]
+            memory_mask: [target sequence length, source sequence length]
+            Note: The maksed positions are filled with float('-inf').
+                  Unmasked positions are filled with float(0.0). Masks ensure that the predictions
+                  for position i depend only on the information before position i.
+            output: [target sequence length, batch size, tgt_vocab]
+            Note: Due to the multi-head attention architecture in the transformer model,
+                  the output sequence length of a transformer is same as the input sequence
+                  (i.e. target) length of the decode.
+        Examples:
+            >>> output = transformer_model(src, tgt, src_mask=src_mask, tgt_mask=tgt_mask)
+        """
+
+        src = self.src_embed(src) * math.sqrt(self.d_model)
+        src = self.pos_encoder(src)
+        tgt = self.tgt_embed(tgt) * math.sqrt(self.d_model)
+        tgt = self.pos_decoder(tgt)
+
+        memory = self.transformer.encoder(src, src_mask)
+        output = self.transformer.decoder(tgt, memory, tgt_mask, memory_mask)
+
+        if self.generator:
+            output = self.generator(output)
+
+        return output
+
+
+# Temporarily leave PositionalEncoding module here. Will be moved somewhere else.
+class PositionalEncoding(nn.Module):
+    r"""Inject some information about the relative or absolute position of the tokens
+        in the sequence. The positional encodings have the same dimension as
+        the embeddings, so that the two can be summed. Here, we use sine and cosine
+        functions of different frequencies.
+    .. math::
+        \text{PosEncoder}(pos, 2i) = sin(pos/10000^(2i/d_model))
+        \text{PosEncoder}(pos, 2i+1) = cos(pos/10000^(2i/d_model))
+        \text{where pos is the word position and i is the embed idx)
+    Args:
+        d_model: the embed dim (required).
+        dropout: the dropout value (default=0.1).
+        max_len: the max. length of the incoming sequence (default=5000).
+    Examples:
+        >>> pos_encoder = PositionalEncoding(d_model)
+    """
+
+    def __init__(self, d_model, dropout=0.1, max_len=5000):
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
```

Comment:
What is the purpose of `-math.log(10000.0)`?

Reply:
It's chosen by the paper "Attention is all you need." It forms a longer enough sin/cos wave length.

Reply:
10000.0 is a random number. It is supposed to be large enough to cover the length of the incoming tgt/src sequences.

Reply:
Can't this be factored out to live outside of torch.exp? First multiplying by a very large negative number and then applying exp probably yields numerical instability.